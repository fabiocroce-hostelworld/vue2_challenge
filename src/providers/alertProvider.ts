import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
