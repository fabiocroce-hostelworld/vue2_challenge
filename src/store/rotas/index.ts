import { getRotaServices } from "@/api/clavaAPI"
import { Rotas } from "@/models/Rotas"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_ROTAS: "GET_ROTAS"
  },
  mutations: {
    STORE_ROTAS: "STORE_ROTAS",
    STORE_ROTA: "STORE_ROTA"
  },
  actions: {
    FETCH_ROTA: "fetchRota",
    GENERATE_ROTA: "generateRota"
  }
}

export const rotasTypes = addNamespacePrefix(types, "rotas")

type RotasState = {
  rotas: Array<Rotas>
}

export default {
  namespaced: true,
  state: (): RotasState => ({
    rotas: [] as Array<Rotas>
  }),
  getters: {
    [types.getters.GET_ROTAS](state: RotasState) {
      return state.rotas
    }
  },
  actions: {
    [types.actions.FETCH_ROTA]: async ({ commit }: { commit: Function }, rotaId: number) => {
      const rota = await getRotaServices().fetchRota(rotaId)

      commit(types.mutations.STORE_ROTA, rota)
    },
    [types.actions.GENERATE_ROTA]: async ({ dispatch }: { dispatch: Function }, date: string) => {
      const rotaId = await getRotaServices().generateRota(date)

      await dispatch(types.actions.FETCH_ROTA, rotaId)
    }
  },
  mutations: {
    [types.mutations.STORE_ROTAS]: (state: RotasState, rotas: Array<Rotas>) => {
      state.rotas = rotas
    },
    [types.mutations.STORE_ROTA]: (state: RotasState, rota: Rotas) => {
      state.rotas.push(rota)
    }
  }
}
