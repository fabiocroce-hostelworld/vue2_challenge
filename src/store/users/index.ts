import { User } from "@/models/User"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_USERS: "getUsers",
    GET_MAPPED_USERS: "getMappedUsers",
    GET_USER_NAME: "getUserName"
  },
  mutations: {
    STORE_USERS: "STORE_USERS"
  },
  actions: {
    FETCH_USERS: "fetchUsers"
  }
}

export const usersTypes = addNamespacePrefix(types, "users")

type UsersState = {
  users: Array<User>
}

export default {
  namespaced: true,
  state: (): UsersState => ({
    users: [] as Array<User>
  }),
  getters: {
    [types.getters.GET_USERS](state: UsersState) {
      return state.users
    },
    [types.getters.GET_MAPPED_USERS](state: UsersState) {
      return Object.assign({}, ...state.users.map(user => ({ [user.userId]: user.user })))
    },
    [types.getters.GET_USER_NAME](_: UsersState, getters: { [x: string]: Array<User> }) {
      return (userId: number) => getters[types.getters.GET_MAPPED_USERS][userId]
    }
  },
  mutations: {
    [types.mutations.STORE_USERS]: (state: UsersState, users: Array<User>) => {
      state.users = users
    }
  }
}
