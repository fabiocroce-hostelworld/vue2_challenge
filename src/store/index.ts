import { getRotasUsersServices } from "@/api/clavaAPI"
import Vue from "vue"
import Vuex, { Store } from "vuex"
import rotas, { rotasTypes } from "@/store/rotas"
import users, { usersTypes } from "@/store/users"

Vue.use(Vuex)

export const rootTypes = {
  actions: {
    FETCH_ROTAS_AND_USERS: "fetchRotasAndUsers"
  }
}

export { rotasTypes, usersTypes }

export const storeConfig = {
  modules: {
    rotas,
    users
  },
  actions: {
    [rootTypes.actions.FETCH_ROTAS_AND_USERS]: async ({ commit }: { commit: Function }) => {
      const { users, rotas } = await getRotasUsersServices().fetchUsersAndRotas()

      commit(rotasTypes.mutations.STORE_ROTAS, rotas)
      commit(usersTypes.mutations.STORE_USERS, users)
    }
  }
}

export default new Store(storeConfig)
