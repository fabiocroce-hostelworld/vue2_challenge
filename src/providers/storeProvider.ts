import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
