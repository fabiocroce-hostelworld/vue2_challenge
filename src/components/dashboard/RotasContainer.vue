<template>
  <v-container fluid>
    <v-row>
      <v-col sm="12" md="3">
        <users-list :users="users" @change="filterUsers($event, selectedUsers)" />
      </v-col>
      <v-col sm="12" md="9">
        <rotas-calendar :selectedUsers="selectedUsers.users" />
        <br />
        <add-rota />
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from "vue"
import { usersTypes } from "@/store"
import { StoreProvider, useStore } from "@/providers/storeProvider"
import { reactive } from "@vue/composition-api"
import RotasCalendar from "@/components/dashboard/RotasCalendar.vue"
import UsersList from "@/components/dashboard/UsersList.vue"
import AddRota from "@/components/dashboard/AddRota.vue"

export default Vue.extend({
  name: "RotasDisplay",
  components: {
    RotasCalendar,
    UsersList,
    AddRota
  },
  setup() {
    const { reactiveGetter } = useStore() as StoreProvider
    const selectedUsers = reactive({ users: new Set() })

    const users = reactiveGetter(usersTypes.getters.GET_USERS)

    const filterUsers = (updatedUsers: Array<number>, actualSelectedUsers: { users: Set<number> }) => {
      actualSelectedUsers.users = new Set([...updatedUsers])
    }

    return {
      users,
      selectedUsers,
      filterUsers
    }
  }
})
</script>
