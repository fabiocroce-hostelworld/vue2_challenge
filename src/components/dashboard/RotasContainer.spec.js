import Vuex from "vuex"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import { storeConfig, usersTypes } from "@/store"
import RotasContainer from "@/components/dashboard/RotasContainer"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"
import { setupStoreProvider, storeRef } from "@/providers/storeProvider"

describe("RotasContainer", () => {
  let input
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()
    store = new Vuex.Store(storeConfig)

    store.commit(usersTypes.mutations.STORE_USERS, rotaUsersMock.users)

    input = {
      localVue,
      provide: {
        [storeRef]: setupStoreProvider(store)
      }
    }
  })

  it("Snapshot test", () => {
    const wrapper = shallowMount(RotasContainer, input)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("when change event emitted expect RotasDisplay to send selectedUsers to RotasCalendar", async () => {
    const wrapper = shallowMount(RotasContainer, input)
    const usersListComponent = wrapper.find("users-list-stub")

    expect(wrapper.find("rotas-calendar-stub").props().selectedUsers).toEqual(new Set())

    await usersListComponent.vm.$emit("change", [1, 2])

    expect(wrapper.find("rotas-calendar-stub").props().selectedUsers).toEqual(new Set([1, 2]))
    expect(usersListComponent.props().users).toBe(rotaUsersMock.users)
  })
})
