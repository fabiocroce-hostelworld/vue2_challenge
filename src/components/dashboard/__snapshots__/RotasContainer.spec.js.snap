// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotasContainer Snapshot test 1`] = `
<v-container-stub tag="div"
                  fluid="true"
>
  <v-row-stub tag="div">
    <v-col-stub sm="12"
                md="3"
                tag="div"
    >
      <users-list-stub users="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]">
      </users-list-stub>
    </v-col-stub>
    <v-col-stub sm="12"
                md="9"
                tag="div"
    >
      <rotas-calendar-stub selectedusers="[object Set]">
      </rotas-calendar-stub>
      <br>
      <add-rota-stub today="2020-12-17">
      </add-rota-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
